File: language/compiler/src/com/kubadziworski/parsing/visitor/statement/IfStatementVisitor.java
package com.kubadziworski.parsing.visitor.statement;

import jiva.parser.JivaBaseVisitor;
import jiva.parser.JivaParser;
import jiva.parser.JivaParser.ExpressionContext;
import jiva.parser.JivaParser.IfStatementContext;
import com.kubadziworski.domain.node.expression.Expression;
import com.kubadziworski.domain.node.statement.IfStatement;
import com.kubadziworski.domain.node.statement.Statement;
import com.kubadziworski.parsing.visitor.expression.ExpressionVisitor;
import org.antlr.v4.runtime.misc.NotNull;

public class IfStatementVisitor extends JivaBaseVisitor<IfStatement> {
    private final StatementVisitor statementVisitor;
    private final ExpressionVisitor expressionVisitor;

    public IfStatementVisitor(StatementVisitor statementVisitor, ExpressionVisitor expressionVisitor) {
        this.statementVisitor = statementVisitor;
        this.expressionVisitor = expressionVisitor;
    }

    @Override
    public IfStatement visitIfStatement(@NotNull IfStatementContext ctx) {
        ExpressionContext conditionalExpressionContext = ctx.expression();
        Expression condition = conditionalExpressionContext.accept(expressionVisitor);
        Statement trueStatement = ctx.trueStatement.accept(statementVisitor);
        if (ctx.falseStatement != null) {
            Statement falseStatement = ctx.falseStatement.accept(statementVisitor);
            return new IfStatement(condition, trueStatement, falseStatement);
        }
        return new IfStatement(condition, trueStatement);
    }
}

File: language/compiler/src/parsing/visitor/statement/IfStatementVisitor.kt
package jiva.parsing.visitor.statement

import jiva.domain.node.statement.IfStatement
import jiva.parsing.visitor.expression.ExpressionVisitor
import jiva.parser.JivaBaseVisitor
import jiva.parser.JivaParser.IfStatementContext

class IfStatementVisitor(
    private val statementVisitor: StatementVisitor,
    private val expressionVisitor: ExpressionVisitor
) : JivaBaseVisitor<IfStatement>() {

    override fun visitIfStatement(ctx: IfStatementContext): IfStatement {
        val conditionalExpressionContext = ctx.expression()
        val condition = conditionalExpressionContext.accept(expressionVisitor)!!
        val trueStatement = ctx.trueStatement.accept(statementVisitor)!!
        if (ctx.falseStatement != null) {
            val falseStatement = ctx.falseStatement.accept(statementVisitor)!!
            return IfStatement(condition, trueStatement, falseStatement)
        }
        return IfStatement(condition, trueStatement)
    }

}